### PR TITLE
Adjust achievement unlock rate to display Hardcore unlock rate

### PR DESCRIPTION
**Pull request by endearall**
This changes an achievement's unlock rate current UI:

183 **(63)** of 3,235
5.66% unlock rate

To:

**63** (183) of 3,235
**1.95%** (5.66) unlock rate

If the achievement has no hardcore unlock, it will default to (just like current behavior):

183 of 3,235
5.66% unlock rate

Showing hardcore unlock rate as a more useful metric for achievement difficulty is a requested feature in:

https://retroachievements.org/viewtopic.php?t=28159
https://retroachievements.org/viewtopic.php?t=28500
https://retroachievements.org/viewtopic.php?t=28575

### CODE REVIEW

**Comment by wescopeland** [comment]
Hi @ladynadiad, thanks for voicing that suggestion.

Respectfully, I believe for the time being that may be too assertive of a change.

At the moment, the code changes in this branch don't remove the visual display of softcore earn rate from achievements (assuming there is one). This branch separates the hardcore and softcore earn rates and designates the hardcore rate as the primary statistic.

If we are to consider visually hiding information altogether, there are a lot of nuances to think through, including:
* Whether our internal analytics data supports making such a change and how this change might steer user behavior (if possible, we want to steer softcore players towards giving hardcore a try).
* Respecting player context (as a hardcore player with no softcore rank, I may never want to see softcore stuff on most pages).
* Respecting user abilities (as a developer, I may actually be interested to see softcore stuff on my sets).
* Respecting user preference (I may still want to see stuff regardless of what the system assumes I want to see).

File: resources/views/components/game/achievements-list/list-item-global-progress.blade.php
[the remark this answers is not in the document]
Comment (endearall) [reply resources/views/components/game/achievements-list/list-item-global-progress.blade.php:1]: Hi @wescopeland, Thank you for your review and detailed feedback! And thank you for spotting the display issue if there is only hardcore progress. I have made the changes accordingly.

**Comment by Jamiras** [comment]
I support this change, only for the fact that it makes the numbers match up with the order of the data in the progress bar.
* Progress bar: [hardcore][softcore][total]
* Message: Hardcore (softcore) of total

However, I'm strongly against trying to eradicate/hide softcore. The driving argument in the forum posts is that having softcore unlocks shown obscures the difficulty of the achievement. As mentioned above, if hiding the gray bar eliminates (roughly) half of the players for a set, the achievement automatically looks twice as difficult as it is "183 (63) of 3,235". would become "63 of 3,235 (1.9%)", when in reality it's probably closer to "63 of 1,832 (3.4%)". 

And if we try to shrink the bar to only account for the hardcore players, then we lose the players that loaded the game but never earned any achievements (notice how most games don't have any achievements with 100% unlock rate?).

-----

The data doesn't actually separate the users into hardcore/softcore buckets. Hardcore is just a tier above softcore. A user who has unlocked an achievement in hardcore has also implicitly unlocked it in softcore. For "183 (63) of 3,235", there's only actually 120 pure softcore unlocks. The bar is stacked:
image

Just like the achievement distribution graph:
image

Which would (presumably) also be affected by hiding the softcore data.

-----

That sort of change would also affect things outside of the web page. In the last year, we've started sending unlock rates to clients to display when the achievement pops. It's using the same math as the webpage where hardcore unlocks have an intrinsically lower percentage because of the portion of users who have only unlocked the achievement in softcore. 

And as has been discussed many times in regards to retro ratio, unlock rate often reflects attrition more than difficulty. A very simple endgame achievement may have an arbitrarily low unlock rate just because people quit playing before they get to it.

Rather than seeing the bar as "softcore players are ruining my stats", I see it as "this many players chose to have help". I still want to know how many people (of _all the people who played the game_) were able to do it without help, and how many were able to do it at all. I haven't tried it recently, but in the early 2000s, I still couldn't beat Battletoads, even with save states.

**Comment by wescopeland** [comment]
> However, I'm strongly against trying to eradicate/hide softcore.

I am also against trying to eradicate/hide softcore.

To clarify, are you also against what I wrote for point 3 in my feedback?
> If there is only hardcore progress, don't show softcore progress.

**Comment by wescopeland** [comment]
Hey @endearall, thanks for working through our arm wrestling! I feel we're getting closer. We may need to do just a bit more on the design front.

Here are two achievements, one with a single hardcore unlock and another with a single softcore unlock:

Screenshot 2024-10-28 at 4 15 38 PM

@Jamiras wrote above:
> What about rearranging the data to highlight hardcore while obscuring softcore?

I think I'd be inclined to go the other way around. As a platform, we've already deemed hardcore as our default operating mode, and this is now codified in [our integration docs](https://github.com/RetroAchievements/rcheevos/wiki/rc_client-integration#enabling-hardcore).

In other words, hardcore should be the implied statistic, and if something needs to be explicitly stated it should be softcore, as that is the non-default. For example:
Screenshot 2024-10-28 at 4 19 46 PM

For full transparency, I should state I'd be most comfortable merging this PR when both Jamiras and I have approved it. Jamiras's input is incredibly important, and for this item I'm not comfortable being the sole approver.

**Comment by endearall** [comment]
Hi @wescopeland, thank you for the feedback once again!

For distinguishing between hardcore only unlock and softcore only unlock, I thought the bold effect, which is consistently used in other cases to represent hardcore numbers, is enough, but yes adding 'softcore' to the label for softcore only unlock case would make it even clearer. I've implemented it as suggested.

**Comment by wescopeland** [comment]
Hey @endearall, IMO things are looking good at SM, MD, and XL now.

However, I'm still dissatisfied with what I'm seeing at XS and LG. Those two breakpoints in particular represent over 33% of traffic to the page, so it's important that we give them special care.

With XS, I am measuring at the resolution of an iPhone SE.

Screenshot 2024-10-29 at 7 51 38 AM

Screenshot 2024-10-29 at 7 51 43 AM

I'm definitely open to ideas, but we'll need to find a way to not have this additional word wrap at these breakpoints.

**Comment by endearall** [comment]
Hi @wescopeland, thank you again for the feedback and spotting a typo for display less than md.

As for the word wrap at xs and lg, I can guess this is the reason why in the first place it was decided to not include both hardcore unlock rates and total unlock rates.

I guess for this PR the only way to not have word wrap in these breakpoints is to choose to display only the hardcore unlock rates.

**Comment by wescopeland** [comment]
At those breakpoints, it may be worth exploring Jamiras's suggestion above. Though you will need to test it on a local environment to be sure.

**Comment by wescopeland** [comment]
Another option that may be worth considering is at those breakpoints if the decimal places actually matter.

eg: Instead of rendering "9.81%" at XS/LG, render "9%". I haven't tested this idea, but it may give us enough space back to prevent the word wrap. On XS, this has the added benefit of making the rows less data dense.

**Comment by endearall** [comment]
Ah, about my sentence:

> I guess for this PR the only way to not have word wrap in these breakpoints is to choose to display only the hardcore unlock rates.

I mean when the display is like this: **1.95%** (5.66%) unlock rate
To conditionally hide/give up the (5.66%) part for certain breakpoints where space is limited
I don't mean to say to display only the hardcore unlock rate for all breakpoints

Plus the 'softcore' word that was requested earlier also needs another consideration.

As for setting up local environment to test it myself, I'll have to think about it. I'm new to web technologies, and not sure if I have the energy to deep dive into it now.

**Review by wescopeland — CHANGES_REQUESTED** [review]
Hi @endearall,

I definitely appreciate your enthusiasm towards addressing this ask from the community. However, I think it may be best to pause here and take a step back.

Making UI changes without being able to actually see the UI in a local development environment is not a sustainable approach. While the browser devtools can indeed be helpful, they don't give you the full picture of how your changes will behave when deployed.

This is slightly on us, so I must apologize. This expectation wasn't made clear in our _CONTRIBUTING.md_ file. [We're updating that document](https://github.com/RetroAchievements/RAWeb/pull/2792/files) to better clarify this expectation.

Since RetroAchievements is a passion project maintained purely by volunteers, we believe in collaborative development where contributors and reviewers each play important roles. Testing changes in a local environment before submission helps volunteer maintainers focus on providing meaningful feedback about your ideas and implementation rather than handling basic testing.

Setting up a local environment will allow you to:
1. See your changes in real-time.
2. Test across different screen sizes properly.
3. Ensure nothing breaks unexpectedly.

Our _README.md_ has documentation for setting up a local environment.

Once you have a local environment running, we can continue iterating on this PR to get it into a state we're all happy with 😊
